clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -157,21 +157,50 @@ public FilterResult filterTo(TopicPartition partition, RecordFilter filter, Byte
         return filterTo(partition, batches(), filter, destinationBuffer, maxRecordBatchSize, decompressionBufferSupplier);
     }
 
-    private static FilterResult filterTo(TopicPartition partition, Iterable<MutableRecordBatch> batches,
-                                         RecordFilter filter, ByteBuffer destinationBuffer, int maxRecordBatchSize,
-                                         BufferSupplier decompressionBufferSupplier) {
+    private static class FilteredBatchesMetadata {
```

Comment:
Can you please explain why we added this new class and only have a subset of fields in it?

Reply:
This class is tracking state for all filtered batches. Earlier, all branches in `filterTo` were cluttered with logic to track state per batch and aggregating that over all the batches we have seen so far. With this class, we now only track per batch state and `FilteredBatchesMetadata` holds the aggregate. Cleaner separation and avoiding code repetition, mostly which also makes it easier to reason about which fields have / have not been initialized in a particular branch.

Two fields have been left out from here: `messagesRead` and `bytesRead`. Both these fields are aggregated for all the data we have seen so far, regardless of whether we end up filtering it or not.

Reply:
Thanks for the explanation. Maybe worth documenting the intent (briefly).

Reply:
I think it might be nicer to make the fields in `FilterResult` mutable and move the functionality there. It's a little annoying to have an additional internal class just for accumulating a subset of the filter stats.